actors/builtin/market/testing.go

```diff
@@ -0,0 +1,229 @@
+package market
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/ipfs/go-cid"
+	"github.com/pkg/errors"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin"
+	"github.com/filecoin-project/specs-actors/v2/actors/util/adt"
+)
+
+type StateSummary struct {
+	ProposalCount        uint64
+	PendingProposalCount uint64
+	DealStateCount       uint64
+	LockTableCount       uint64
+	DealOpEpochCount     uint64
+	DealOpCount          uint64
+}
+
+// Checks internal invariants of market state.
+func CheckStateInvariants(st *State, store adt.Store) (*StateSummary, *builtin.MessageAccumulator, error) {
+	acc := &builtin.MessageAccumulator{}
+
+	acc.Require(
+		st.TotalClientLockedCollateral.GreaterThanEqual(big.Zero()),
+		"negative total client locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalProviderLockedCollateral.GreaterThanEqual(big.Zero()),
+		"negative total provider locked collateral: %v", st.TotalClientLockedCollateral)
+
+	acc.Require(
+		st.TotalClientStorageFee.GreaterThanEqual(big.Zero()),
+		"negative total client storage fee: %v", st.TotalClientLockedCollateral)
+
+	//
+	// Proposals
+	//
+
+	allIDs := make(map[abi.DealID]struct{})
+	proposalCids := make(map[cid.Cid]struct{})
+	maxDealID := int64(-1)
+	proposalCount := uint64(0)
+
+	proposals, err := adt.AsArray(store, st.Proposals)
+	if err != nil {
+		return nil, acc, err
+	}
+	var proposal DealProposal
+	err = proposals.ForEach(&proposal, func(dealID int64) error {
+		allIDs[abi.DealID(dealID)] = struct{}{}
+
+		pcid, err := proposal.Cid()
+		if err != nil {
+			return err
+		}
+
+		// keep some state
+		proposalCids[pcid] = struct{}{}
+		if dealID > maxDealID {
+			maxDealID = dealID
+		}
+		proposalCount++
+
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	// next id should be higher than any existing deal
+	acc.Require(int64(st.NextID) > maxDealID, "next id, %d, is not greater than highest id in proposals, %d", st.NextID, maxDealID)
+
+	//
+	// Deal States
+	//
+
+	dealStateCount := uint64(0)
+	dealStates, err := adt.AsArray(store, st.States)
+	if err != nil {
+		return nil, acc, err
+	}
+	var dealState DealState
+	err = dealStates.ForEach(&dealState, func(dealID int64) error {
+		acc.Require(
+			dealState.SectorStartEpoch >= 0,
+			"deal state start epoch undefined %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.LastUpdatedEpoch == epochUndefined || dealState.LastUpdatedEpoch >= dealState.SectorStartEpoch,
+			"deal state last updated before sector start %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.SlashEpoch == epochUndefined || dealState.SlashEpoch >= dealState.SectorStartEpoch,
+			"deal state slashed before sector start %d: %v", dealID, dealState)
```

Comment:
There are a couple more assertions in `updatePendingDealState` that we could add here (some in terms of the current epoch, which you can add as a parameter)